RHDEVDOCS-6872: Content creation for Argo CD Agent install (#101499)
openshift/openshift-docs: modules/gitops-setting-up-a-spoke-cluster-environment.adoc

:_mod-docs-content-type: PROCEDURE
[id="gitops-setting-up-a-spoke-cluster-environment_{context}"]
= Setting up a spoke cluster environment

After you set up the Principal component, you can connect it with one or more Agents.

.Prerequisites

* The Principal setup is complete and running.
* You have access to both the Principal and Agent clusters.
* The `argocd-agentctl` CLI tool is installed and accessible from your environment.
* The Helm CLI is installed and configured.

.Procedure

. Create the Agent secret on the Principal cluster.
+
To create a secret on the Principal cluster that contains configuration details for the Agent, run the following command:
+
[source,terminal]
----
$ argocd-agentctl agent create "<AGENT_NAME>" \
  --principal-context "<PRINCIPAL_CONTEXT>" \
  --principal-namespace "<PRINCIPAL_NAMESPACE>" \
  --resource-proxy-server "<PRINCIPAL_RESOURCE_PROXY_URL>"
----
+
where:
+
`<AGENT_NAME>`:: A unique name used to identify each Agent managed by the Principal. The Agent CLI uses this value to create the Argo CD cluster secret (`cluster-<agent-name>`) and to construct the server URL.

> `<resource-proxy-server>`:: Specifies the URL of the Principal resource-proxy, for example, `openshift-gitops-agent-principal-resource-proxy:9090`, which Argo CD uses to access live resources in the Agent's cluster. In this model, Argo CD does not connect directly to the Agent cluster's API server; rather, it sends requests to the Principal's `resource-proxy`, which then forwards them to the appropriate Agent via gRPC.

Reviewer: Please replace `openshift-gitops-agent-principal-resource-proxy` with `(ArgoCD Instance Name)-agent-principal-resource-proxy`

:_mod-docs-content-type: PROCEDURE
[id="gitops-setting-up-a-spoke-cluster-environment_{context}"]
= Setting up a spoke cluster environment

After you set up the Principal component, you can connect it with one or more Agents.

.Prerequisites

* The Principal setup is complete and running.
* You have access to both the Principal and Agent clusters.
* The `argocd-agentctl` CLI tool is installed and accessible from your environment.
* The Helm CLI is installed and configured.

.Procedure

. Create the Agent secret on the Principal cluster.
+
To create a secret on the Principal cluster that contains configuration details for the Agent, run the following command:
+
[source,terminal]
----
$ argocd-agentctl agent create "<AGENT_NAME>" \
  --principal-context "<PRINCIPAL_CONTEXT>" \
  --principal-namespace "<PRINCIPAL_NAMESPACE>" \
  --resource-proxy-server "<PRINCIPAL_RESOURCE_PROXY_URL>"
----
+
where:
+
`<AGENT_NAME>`:: A unique name used to identify each Agent managed by the Principal. The Agent CLI uses this value to create the Argo CD cluster secret (`cluster-<agent-name>`) and to construct the server URL.

> `<resource-proxy-server>`:: Specifies the URL of the Principal resource-proxy, for example, `(ArgoCD Instance Name)-agent-principal-resource-proxy:9090`, which Argo CD uses to access live resources in the Agent's cluster. In this model, Argo CD does not connect directly to the Agent cluster's API server; rather, it sends requests to the Principal's `resource-proxy`, which then forwards them to the appropriate Agent via gRPC.

Reviewer 2: ```suggestion
`<resource-proxy-server>`:: Specifies the URL of the Principal resource-proxy, for example, `<ArgoCD Instance Name>-agent-principal-resource-proxy:9090`, which Argo CD uses to access live resources in the Agent's cluster. In this model, Argo CD does not connect directly to the Agent cluster's API server; rather, it sends requests to the Principal's `resource-proxy`, which then forwards them to the appropriate Agent via gRPC.
```

Author: Fixed.

`<agent-namespace>`:: Specifies the namespace where the Agent component is installed.
`<principal-namespace>`:: Specifies the namespace where the Principal component is installed.
+
This command creates the `argocd-agent-client-tls` secret, generates the Agent's client certificate and private key, and signs the certificate using the Principal CA.

. Verify the required secrets.
+
To verify that all required secrets are created in the Agent namespace, run the following command:
+
[source,terminal]
----
$ oc get secrets -n "<AGENT_NAMESPACE>" --context "<AGENT_CONTEXT>" | grep agent
----
+
As a result, the following output is displayed:
+
[source,terminal]
----
NAME                              TYPE
argocd-agent-ca                   kubernetes.io/tls
argocd-agent-principal-tls        kubernetes.io/tls
argocd-agent-resource-proxy-tls   kubernetes.io/tls
argocd-agent-jwt                  Opaque
----

. Deploy the Argo CD instance on the Agent cluster.
+
To deploy an Argo CD instance on the Agent cluster, create an Argo CD custom resource similar to the following example:
+
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:

>   name: openshift-gitops

Reviewer 2: Are we recommending the default instance?

Reviewer 2: @Dhruv-Soni11 ignore this comment

// Module included in the following assembly:
//
// * argocd-agent-installation/gitops-setting-up-a-spoke-cluster-environment.adoc

:_mod-docs-content-type: PROCEDURE
[id="gitops-setting-up-a-spoke-cluster-environment_{context}"]
= Setting up a spoke cluster environment

> After you set up the Principal component, you can connect it with one or more Agents.

Author: Fixed.

.Prerequisites

* The Principal setup is complete and running.
* You have access to both the Principal and Agent clusters.
* The `argocd-agentctl` CLI tool is installed and accessible from your environment.
* The `helm` CLI is installed and configured. Ensure that the `helm` CLI version is later than v3.8.0.

.Procedure

. Create the Agent secret on the Principal cluster.
+
To create a secret on the Principal cluster that contains configuration details for the Agent, run the following command:
+
[source,terminal]
----
$ argocd-agentctl agent create "<AGENT_NAME>" \
  --principal-context "<PRINCIPAL_CONTEXT>" \
  --principal-namespace "<PRINCIPAL_NAMESPACE>" \

>   --resource-proxy-server "<PRINCIPAL_RESOURCE_PROXY_URL>"

Reviewer 3: lowercase

Author: Fixed.

----
+
where:
+
`<AGENT_NAME>`:: A unique name used to identify each Agent managed by the Principal. The Agent CLI uses this value to create the Argo CD cluster secret (`cluster-<agent-name>`) and to construct the server URL.
`<resource-proxy-server>`:: Specifies the URL of the Principal resource-proxy, for example, `<ArgoCD Instance Name>-agent-principal-resource-proxy:9090`, which Argo CD uses to access live resources in the Agent's cluster. In this model, Argo CD does not connect directly to the Agent cluster's API server; rather, it sends requests to the Principal's `resource-proxy`, which then forwards them to the appropriate Agent via gRPC.
`<principal-namespace>`:: Specifies the namespace where the Principal component is installed.
`<principal-context>`:: Specifies the cluster context of the Principal component.
+
**Additional options:**
`<label, l>`:: Adds optional metadata labels for the Agent.
+
This command generates a client certificate signed by the Principal CA, creates an Argo CD cluster secret with the certificate data and sets up credentials for resource proxy access.

. Copy the CA certificate to the Agent cluster.
+

> To copy the Principal CA certificate to the Agent cluster for validation, run the following command:

Reviewer 3: ```suggestion
. Copy the CA certificate to the Agent cluster by running the following command:
```

+
[source,terminal]
----
$ argocd-agentctl pki propagate \
    --agent-context "<AGENT_CONTEXT>" \
    --principal-context "<PRINCIPAL_CONTEXT>" \
    --principal-namespace "<PRINCIPAL_NAMESPACE>" \
    --agent-namespace "<AGENT_NAMESPACE>"

> ----

Reviewer 3: lowercase

Author: Fixed.

+
where:
+
`<principal-namespace>`:: Specifies the namespace where the Principal component is installed.
`<agent-namespace>`:: Specifies the namespace where the Agent component is installed.
`<principal-context>`:: Specifies the kubeconfig context of the Principal (hub) cluster.
`<agent-context>`:: Specifies the kubeconfig context of the Agent (workload) cluster.
+
**Additional options:**

> `<force, f>`:: Forces regeneration of the PKI if it already exists on the target cluster.

Reviewer 3: need to be variable described, not the keys

+
This command copies the CA certificate from the Principal cluster and creates the `argocd-agent-ca` secret in the Agent cluster.

. Generate a client certificate for the Agent.
+

> To generate a client certificate and private key for the Agent, run the following command:

Author: Fixed.

+
[source,terminal]
----
$ argocd-agentctl pki issue agent "<AGENT_NAME>" \
    --principal-context "<PRINCIPAL_CONTEXT>" \
    --agent-context "<AGENT_CONTEXT>" \
    --agent-namespace "<AGENT_NAMESPACE>" \
    --principal-namespace "<PRINCIPAL_NAMESPACE>"
----
+
where:
+
`<principal-context>`:: Specifies the kubeconfig context of the Principal (hub) cluster.
`<agent-context>`:: Specifies the kubeconfig context of the Agent (workload) cluster.
`<agent-namespace>`:: Specifies the namespace where the Agent component is installed.
`<principal-namespace>`:: Specifies the namespace where the Principal component is installed.
+
**Additional options:**

> `<upsert, u>`:: Updates the existing JWT key if one already exists in the Agent namespace.

Reviewer 3: same comments here as previous ones

+
This command creates the `argocd-agent-client-tls` secret, generates the Agent's client certificate and private key, and signs the certificate using the Principal CA.

. Verify the required secrets.
+
To verify that all required secrets are created in the Agent namespace, run the following command:
+
[source,terminal]
----

> $ oc get secrets -n "<AGENT_NAMESPACE>" --context "<AGENT_CONTEXT>" | grep agent

Author: Fixed.

----
+
As a result, the following output is displayed:
+
[source,terminal]
----
NAME                              TYPE
argocd-agent-ca                   kubernetes.io/tls
argocd-agent-principal-tls        kubernetes.io/tls
argocd-agent-resource-proxy-tls   kubernetes.io/tls
argocd-agent-jwt                  Opaque
----

. Deploy the Argo CD instance on the Agent cluster.
+

> To deploy an Argo CD instance on the Agent cluster, create an Argo CD custom resource similar to the following example:

Reviewer 3: ```suggestion
. Deploy an Argo CD instance on the Agent cluster by creating an Argo CD CR similar to the following example:
```

+
[source,yaml]
----
apiVersion: argoproj.io/v1beta1
kind: ArgoCD
metadata:
  name: openshift-gitops
spec:
  server:
    enabled: false
----
+
Apply the configuration by running the following command:
+
[source,terminal]
----
$ oc apply -f argocd.yaml -n "<AGENT_NAMESPACE>" --context "<AGENT_CONTEXT>"
----
+
To verify that the Argo CD instance is created, run the following command:

> +

Author: Fixed.
